### ISSUE TITLE
Add support for HTML5 Notification API

### ISSUE THREAD

**Issue by dstaley** [MicrosoftEdge/WebViewFeedback#308]
It would be really nice if WebView2 had appropriate APIs to handle granting notification permissions and displaying notifications from web content that uses the standard Notifications API.

**Comment by pagoe-msft** [comment]
@dstaley - thanks for the feature request! This something we have in the works, I will return to this thread when we have more clarity on when this feature will be completed.

**Comment by peakpeak-github** [comment]
> @dstaley - thanks for the feature request! This something we have in the works, I will return to this thread when we have more clarity on when this feature will be completed.

This was 3 months ago. Any news on the issue?

**Comment by yuvdev** [comment]
Still waiting

**Comment by champnic** [comment]
Sorry for the delay!

Would the [PermissionRequested ](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/0-9-622/icorewebview2#add_permissionrequested) event work for you scenario?

**Comment by yuvdev** [comment]
PermissionRequested not working at all, i implemented client from that tutorial:
https://codelabs.developers.google.com/codelabs/push-notifications/#0
all works fine in Edge (86.0.622.38)
but i got 
"DOMException: Registration failed - permission denied" when i tried to enable push notifications.
"TypeError: Failed to execute 'showNotification' on 'ServiceWorkerRegistration': No notification permission has been granted for this origin." when i try to push message.

I'm using 
<package id="Microsoft.Web.WebView2" version="0.9.579-prerelease" targetFramework="net472" />
for WPF

Version 0.9.628-prerelease not work at all.

**Comment by champnic** [comment]
Are you not getting the PermissionRequested event firing on the WebView? Can you link what part of that tutorial you are using for permission registration?

**Comment by tomhoh** [comment]
Any updates or timeline?

**Comment by minecraftchest1** [comment]
I don't know if this is important to this, but I can't get the permissionRequested even to fire using C# and WPF. It builds fine but doesn't run so I can't see what difference that makes. No errors are shown during runtime.

**Comment by tomhoh** [comment]
Same here.  The permissonRequested never fires.

**Comment by Deutschi** [comment]
If I understand correctly, you basically removed this feature with version 0.8.314 https://docs.microsoft.com/en-us/microsoft-edge/webview2/release-notes#08314

> Updated Notification Change Behavior so WebView2 automatically rejects notification permission requests prompted by web content hosted in the WebView.

This issue is now close to its 1-year anniversary. Are there any updates?

**Comment by DogFoxX** [comment]
Anyone getting this to work? I've tried
`webView21.CoreWebView2.PermissionRequested += CoreWebView2_PermissionRequested;

private void CoreWebView2_PermissionRequested(object sender, Microsoft.Web.WebView2.Core.CoreWebView2PermissionRequestedEventArgs e)
{
    e.State = CoreWebView2PermissionState.Allow;
}`

But when I try to run my app, it throws an exception: Object reference not set to an instance of an object.

**Comment by tomhoh** [comment]
Nope. The PermissionRequested event does not fire in .net. Works on Win32. Waiting for bug to be addressed.

**Comment by DogFoxX** [comment]
Got it working with this code I got from another issue, can't seem to find it again though:

```
async private void InitializeWebView2()
        {
            await webView21.EnsureCoreWebView2Async();
            webView21.CoreWebView2.PermissionRequested += CoreWebView2_PermissionRequested;
        }

        private void CoreWebView2_PermissionRequested(object sender, CoreWebView2PermissionRequestedEventArgs e)
        {
            Debug.WriteLine(e.PermissionKind.ToString());
            e.State = CoreWebView2PermissionState.Allow;
        }
```

The event seems to fire now for any Permission Request, all except for notifications. As stated in die Documentation, Notifications are **auto-rejected and no event is raised for it.**
[](https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2permissionkind?view=webview2-dotnet-1.0.864.35#fields)


Notifications | 4 | Indicates permission to send web notifications. This permission request is currently auto-rejected and no event is raised for it.
-- | -- | --




It's a shame this hasn't been addressed yet.

**Comment by DominicFarrington** [comment]
Also waiting on this feature; any timeline on this? A timeline would be helpful because it would help inform whether or not to implement a custom work around.

Thanks

**Comment by champnic** [comment]
We don't currently have this work scheduled within the next quarter.

**Comment by koenvd** [comment]
This would also be very useful for us since currently we are using a service worker to support push notifications in our PWA.

Also not sure what would be the best custom workaround here for a WinUI3.0 app? Using the push notifications API as being used by UWP apps? (Not sure that API is available for desktop apps). Or is there another alternative that we could use?

**Comment by koenvd** [comment]
Just wondering if this still is on the radar @champnic ? Would still be awesome to have this supported ...

**Comment by champnic** [comment]
It's on the radar. It's not scheduled for this quarter, but if it's priority on our backlog doesn't change, then I think we would begin work next quarter. The current workaround is indeed to use the existing Windows Notifications APIs from the host app.

**Comment by koenvd** [comment]
Thanks for update @champnic . Much appreciated.

**Comment by suzanneruigrok** [comment]
@champnic Wondering if this has this been picked up in Q2? Thanks a lot in advance for your reply :)

**Comment by datvm** [comment]
The even isn't fired for me as well (on WebUI3), I am requesting Camera and Microphone using `navigator.mediaDevices.getUserMedia`. I blocked it on the first time it asked and now I can't seem to find anyway to change it again.

**Comment by champnic** [comment]
@maurawinstanley Do you know if one of the "ClearBrowsingData" options works for permissions state?

**Comment by datvm** [comment]
Hi is there any update on this? Any workaround I can use in the meantime?

**Comment by Kidades** [comment]
Any update on this?

**Comment by wilsonmfg** [comment]
also looking for updates. it's not clear how one would use the Windows Notification API unless all the code is under ones control.

for example, if I want to embed the web based Teams client in my application, I have no options that I can figure out.

**Comment by jasonstephen15** [comment]
Hi there! The WebView2 team is actively spec'ing out this feature request right now. The major requirements I've noted down based on the comments in chat are:

1. Developers want permission event to fire for notifications.
2. Developers want a way to enable changing permissions after user selects the first time.
3. Developers want some level of control/customization of the notification coming from the web. (Ex. changing text, or picture, or callback link).

I will also go ahead and suggest the following requirements, but please correct me if this doesn't make sense:

1. The Edge permission UI should not have the website name, but rather the parent app name or simply a generic prompt.
![image](https://user-images.githubusercontent.com/22718545/188245658-1d5fe809-44bd-4c7b-97a1-1b3c9ddb92ce.png)

2. The native notification in Windows should not have any Edge/WebView branding.

--- Other than that, I do have a couple questions I was hoping to get customer input on:
1. If a notification is shown to the end user, but the app/webview2 is closed, what is the expected behavior? I might offer that the app provide a protocol/uri along with the notification for callback, but if this isn't provided what should the default case be? Open in Edge?

Looking for any thoughts and further input! Thanks in advance!

**Comment by wilsonmfg** [comment]
I hadn't considered that the notifications would work when the app is closed.  If there's a way to provide a a method to open the app that would be great, I already use a custom protocol handler so that fits well.  I see plenty of windows notifications now that clicking them simply dismisses them, so I think that would be acceptable as a fallback.

**Comment by OkkioXavier** [comment]
@jasonstephen15 I want to be able to enable notifications without user interaction.

**Comment by jasonstephen15** [comment]
@OkkioXavier - what do you mean by that? do you mind expanding on your app scenario that requires notifications? Are you talking about push/background notifications through [WNS](https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/windows-push-notification-services--wns--overview)? Most of the stuff I've outlined is to get e2e support for HTML5 Notifications API.

**Comment by OkkioXavier** [comment]
@jasonstephen15 thanks for your reply.

Some more details:

We're implementing a line of business application which is deployed to enterprise users, it's important for our use case that user permission is not required for push notifications. 

The application which we are replacing is built in WinForms and uses the [NotifyIcon ](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.notifyicon?redirectedfrom=MSDN&view=windowsdesktop-6.0) class to display notifications to users. User interaction is not required to display a NotifyIcon.

WebView2 already implements support for the PermissionRequested event for other HTML5 functionality such as ClipboardRead. For those events you can programmatically grant/deny permission. When permission is granted programmatically no permissions dialog is displayed to the user. Above I was asking that the existing pattern is applied to the PermissionRequested event when raised for push notifications.

Sample code for ClipboardRead:

 ```
view.CoreWebView2.PermissionRequested += (_, args) =>
{
  var uri = new Uri(args.Uri);
  
  if (args.PermissionKind == CoreWebView2PermissionKind.ClipboardRead && uri.IsTrusted(allowLocalhost))
  {
	  args.State = CoreWebView2PermissionState.Allow;
  }
  else
  {
	  args.State = CoreWebView2PermissionState.Default;
  }
};		
```

Thanks! Let me know if you need any further details

**Comment by jasonstephen15** [comment]
@OkkioXavier - right, in earlier message I noted `Developers want permission event to fire for notifications.` I think this should be all you need to light up your scenario. Let me know if you think otherwise. Just for absolute certainty when you say 'push notifications', you mean HTML5 Notifications being raised on the native side right?

**Comment by DominicFarrington** [comment]
@jasonstephen15 Yep, exactly. I was just clarifying that when the event fires we want the existing permissions grant functionality to be maintained. Thanks for following up.

**Comment by grabinat** [comment]
@jasonstephen15 
> The native notification in Windows should not have any Edge/WebView branding.
This sounds like it is planned, that Edge/WebView itself will create the toast notifications showed to the user. Is this optional?

For my use case it would be important, that the WebView is providing the HTML5 Notifications payload to the app without showing any notification itself. The payload would then be processed by the app code, and the toast notification would then be created by the app itself.

**Comment by wilsonmfg** [comment]
if the event allowed updating the payload and the option to set a Handled property like keypress events do that would tell the webview to not show the notification, that would be amazing.

**Comment by RyanGelinas** [comment]
Looks like's it been a few months since the last update, is this still on the roadmap?  I also have an enterprise application this would be key to.  I'd need to be able to programmatically allow the push and then handle notifications to the user in my code, since it will be primary operating on devices that use a custom shell, which doesn't support native windows notifications.

**Comment by champnic** [comment]
I believe we have begun work on this feature - @peiche-jessica should have the latest.

**Comment by LafallaMohamed** [comment]
Hello, Is there any updates regarding this feature ?

**Comment by peiche-jessica** [comment]
It should be available in the next SDK release in early June. Thanks.

**Comment by amilich** [comment]
> It should be available in the next SDK release in early June. Thanks.

just writing once more, let me know if it was released

**Comment by novac42** [comment]
Hi @amilich It's released as experimental API now: please check out https://learn.microsoft.com/en-us/microsoft-edge/webview2/release-notes?tabs=dotnetcsharp#experimental-features

**Comment by minecraftchest1** [comment]
This has taken way to long to be implemented. Will not be using, and I will tell everyone I know not to use.

**Comment by Kidades** [comment]
> This has taken way to long to be implemented. Will not be using, and I will tell everyone I know not to use.

It's only a game, you no heff to be mad.

**Comment by ddart2** [comment]
> Hi @amilich It's released as experimental API now: please check out https://learn.microsoft.com/en-us/microsoft-edge/webview2/release-notes?tabs=dotnetcsharp#experimental-features

Still not working with 1.0.1988-prerelease and 1.0.1901 release.
After setting the permission state of webview2 as 'allow' (=1),
when I try to run the 'subscribe' function of pushManager, it reports an error with the following message:

"subscribe error DOMException: Registration failed - push service error."

My push subscription code works well with Chrome and Edge, but it doesn't work with webview2.

**Comment by champnic** [comment]
@ddart2 What runtime version are you using? You'll likely need to use Edge Canary/Dev for this to work:
https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/set-preview-channel

**Comment by ddart2** [comment]
> @ddart2 What runtime version are you using? You'll likely need to use Edge Canary/Dev for this to work: https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/set-preview-channel

I am using and set enviroment as 117.0.2045.7 edge canary.
it works with new experimental interface except 'web push subscription'.
Should I set permission state through deferral method of args, at the different thread? 
Is there anyone who has succeeded in subscribing to push notifications?
thanks

**Comment by YoHanGLoot** [comment]
Anyone still with this issue? It works for Edge v 118, but not for WebView2

**Comment by ngocvd** [comment]
Still now, it is not fired.

**Comment by grabinat** [comment]
In WinUI3 WebNotifcations are still not available.
I also just tested the WinAppSDK 1.6 experimental1. But it also does not include this feature.
Is there any ETA when this will be rollout to WinUI3 - especially as WinUI3 is now the recommended native UI platform?
